Sources/WebService/AuthenticatedWebService/AuthenticatedWebService.swift

```diff
@@ -56,84 +56,83 @@ open class AuthenticatedWebService: WebService {
+    func appendTokenAndExecute(accessToken: AccessToken) -> AnyPublisher<Void, Error> {
+      urlRequest.setValue(self.configuration.authorizationHeaderScheme.rawValue + accessToken, forHTTPHeaderField: "Authorization")
+      return super.execute(urlRequest: urlRequest)
+        .subscribe(on: DispatchQueue.global())
+        .eraseToAnyPublisher()
     }
 
-    guard let accessToken = currentAccessToken else {
-      return Fail<Void, Error>(error: NetworkError.unauthorized).eraseToAnyPublisher()
+    return Deferred {
+      self.tokenProvider.accessToken
+        .compactMap { $0 }
+        .setFailureType(to: Error.self)
+        .flatMap { accessToken -> AnyPublisher<Void, Error> in
+          return appendTokenAndExecute(accessToken: accessToken)
+        }.eraseToAnyPublisher()
+    }
+    .timeout(10, scheduler: DispatchQueue.main, customError: { NetworkError.timeout })
+    .catch { [weak self] error -> AnyPublisher<Void, Error> in
+      guard let self = self else {
+        return Fail<Void, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
+      }
+
+      if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
+        return self.retrySynchronizedTokenRefresh()
+          .flatMap {
+            appendTokenAndExecute(accessToken: $0)
+        }.eraseToAnyPublisher()
+      }
+      return Fail<Void, Error>(error: error).eraseToAnyPublisher()
     }
-    
-    urlRequest.setValue(self.configuration.authorizationHeaderScheme.rawValue + accessToken, forHTTPHeaderField: "Authorization")
-    
-    return super.execute(urlRequest: urlRequest)
-      .catch { [weak self] error -> AnyPublisher<Void, Error> in
-        guard let self = self else {
-          return Fail<Void, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
-        }
-
-        if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
-          self.retrySynchronizedTokenRefresh()
-
-          return self.execute(urlRequest: urlRequest)
-            .delay(for: 0.2, scheduler: self.authenticationQueue)
-            .eraseToAnyPublisher()
-        }
-        return Fail<Void, Error>(error: error).eraseToAnyPublisher()
-    }.eraseToAnyPublisher()
+    .receive(on: DispatchQueue.main)
+    .eraseToAnyPublisher()
   }
 
-  private func retrySynchronizedTokenRefresh() {
-    let dispatchGroup = DispatchGroup()
-    dispatchGroup.enter()
-
-    authenticationQueue.sync(flags: .barrier) {
-      self.tokenProvider.invalidateAccessToken()
-      self.tokenProvider.reissueAccessToken()
-        .sink(receiveCompletion: { _ in
-          dispatchGroup.leave()
-        },
-              receiveValue: { _ in })
-        .store(in: &self.subscriptions)
-      dispatchGroup.wait()
-    }
+  private func retrySynchronizedTokenRefresh() -> AnyPublisher<AccessToken, Error> {
+    tokenProvider.invalidateAccessToken()
```

Comment:
this should be deffered

Reply:
the consumer of this publisher is already deferred